### PR TITLE
New tool Hapcut2

### PR DESCRIPTION
**Pull request by neoformit**
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [x] - This PR adds a new tool or tool collection

This is already published on TS and installed on Galaxy AU - just pushing to tools-iuc for long-term maintenance.
I've added write access to IUC on the TS repo, so this should transfer smoothly according to past experience!

It fails linting based on the conditional semantics, but nothing is required for these two options (`pacbio` and `ont` require a reference genome):

```
Applying linter inputs... WARNING
.. WARNING: Conditional [optimization] no <when /> block found for select option 'default'
.. WARNING: Conditional [optimization] no <when /> block found for select option 'hic'
```

### CODE REVIEW

File: tools/hapcut2/hapcut2.xml [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+## Create haplotype.out and haplotype.out.phased.VCF
+&& HAPCUT2 --fragments frags.dat --VCF input.vcf --output haplotype.out
+#if $optimization.choice == 'hic':
+  --HiC 1
+#end if
+
+  ]]></command>
+
+  <inputs>
+    <param name="input_bam" type="data" format="bam" label="Input BAM file"/>
```
Comment (bernt-matthias) [inline tools/hapcut2/hapcut2.xml:53]: Sorting is needed for the ban file?

File: tools/hapcut2/hapcut2.xml [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+      <param name="choice" type="select" display="radio" label="Optimization">
+        <option value="default" selected="true">Default</option>
+        <option value="pacbio">Pacbio</option>
+        <option value="ont">Oxford Nanopore</option>
+        <option value="hic">Hi-C</option>
+      </param>
+
+      <when value="pacbio">
+        <param name="reference_fasta" type="data" format="fasta"
+          label="Reference genome fasta file"
```
Comment (bernt-matthias) [inline tools/hapcut2/hapcut2.xml:67]: Dies the tool understand ziped fasta?

File: tools/hapcut2/hapcut2.xml [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+      help="Output fragments collected by extractHAIRS"
+    />
+
+    <section name="advanced" title="Advanced parameters">
+      <param name="maxIS" type="integer" label="Maximum insert size"
+        optional="true" value="1000"
+        help="Maximum insert size for a paired-end read to be considered as a single fragment for phasing."
+      />
+
+      <param name="minIS" type="integer" label="Minimum insert size"
```
Comment (bernt-matthias) [inline tools/hapcut2/hapcut2.xml:94]: Ate there're reasonable min and max values?

File: tools/hapcut2/hapcut2.xml [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+    <!-- Oxford nanopore optimization with ref genome -->
+    <test expect_num_outputs="3">
+      <param name="input_bam" ftype="bam" value="input.bam"/>
+      <param name="input_vcf" ftype="vcf" value="input.vcf"/>
+      <param name="output_fragments" value="1"/>
+      <param name="output_phased" value="1"/>
+      <param name="optimization" value="ont"/>
+      <param name="reference_fasta" value="ref.fasta"/>
+      <output name="frags" ftype="txt" file="output_frag.dat"/>
```
Comment (bernt-matthias) [inline tools/hapcut2/hapcut2.xml:163]: This test data is empty...

File: tools/hapcut2/hapcut2.xml (hunk 2 above)
Comment (bgruening) [reply tools/hapcut2/hapcut2.xml:67]: If those are classical reference genomes, it should use this here: https://github.com/galaxyproject/tools-iuc/blob/9ae0037c379b589aa8e0f825820bb4a7c969cdb2/tools/happy/hap.py.xml#L32

File: tools/hapcut2/hapcut2.xml (hunk 3 above)
Comment (bgruening) [reply tools/hapcut2/hapcut2.xml:94]: Can you also consider to use argument="--minIS" here and elsewhere.

File: tools/hapcut2/hapcut2.xml (hunk 4 above)
Comment (neoformit) [reply tools/hapcut2/hapcut2.xml:163]: Yes, some of the outputs are empty when the input data are so small.

File: tools/hapcut2/hapcut2.xml (hunk 1 above)
Comment (neoformit) [reply tools/hapcut2/hapcut2.xml:53]: Not sure what you mean here, can you explain please?

Comment (neoformit) [reply tools/hapcut2/hapcut2.xml]: Oh the BAM should already be sorted? I've added `help` for this, but should we add this to the tool as a pre-processing step (with samtools etc)?

Comment (bernt-matthias) [reply tools/hapcut2/hapcut2.xml]: I think it's fine as is. Users can sort on their own using another tool, e.g. samtools sort

File: tools/hapcut2/hapcut2.xml (hunk 3 above)
Comment (bernt-matthias) [reply tools/hapcut2/hapcut2.xml]: This is fine. Thanks

**Review by bernt-matthias — APPROVED** [review]

**Comment by bernt-matthias** [comment]
Ready?

File: tools/hapcut2/hapcut2.xml [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+      <output name="haplotype_phased" ftype="vcf" file="output_haplotype.out.phased.vcf"/>
+    </test>
+
+    <!-- Pacbio optimization with ref genome -->
+    <test expect_num_outputs="3">
+      <param name="input_bam" ftype="bam" value="input.bam"/>
+      <param name="input_vcf" ftype="vcf" value="input.vcf"/>
+      <param name="output_fragments" value="1"/>
+      <param name="output_phased" value="1"/>
+      <output name="frags" ftype="txt" file="output_frag.dat"/>
```
Comment (bgruening) [inline tools/hapcut2/hapcut2.xml:188]: Outputs should move below imho to improve readability. Should one of these test only have 2 or 1 output, so the filters are tested?

File: tools/hapcut2/.shed.yml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+categories:
+  - Assembly
+description: "Robust and accurate haplotype assembly for diverse sequencing technologies"
+homepage_url: https://github.com/vibansal/HapCUT2
+long_description: |
+  HapCUT2 is a maximum-likelihood-based tool for assembling haplotypes from DNA sequence reads, designed to "just work" with excellent speed and accuracy. We found that previously described haplotype assembly methods are specialized for specific read technologies or protocols, with slow or inaccurate performance on others. With this in mind, HapCUT2 is designed for speed and accuracy across diverse sequencing technologies.
+name: hapcut2
+owner: galaxy-australia
```
[the remark this answers is not in the document]
Comment (neoformit) [reply tools/hapcut2/.shed.yml:8]: It's a migration - permissions already sorted.

File: tools/hapcut2/hapcut2.xml (hunk 5 above)
Comment (neoformit) [reply tools/hapcut2/hapcut2.xml]: Done!

**Comment by bgruening** [comment]
Thanks @neoformit

**Pull request merged**